lib/std/Build/CompileStep.zig

```diff
@@ -1031,22 +1036,110 @@ fn linkLibraryOrObject(self: *CompileStep, other: *CompileStep) void {
 fn appendModuleArgs(
     cs: *CompileStep,
     zig_args: *ArrayList([]const u8),
-    name: []const u8,
-    module: *Module,
 ) error{OutOfMemory}!void {
-    try zig_args.append("--pkg-begin");
-    try zig_args.append(name);
-    try zig_args.append(module.builder.pathFromRoot(module.source_file.getPath(module.builder)));
+    // First, traverse the whole dependency graph and give every module a unique name, ideally one
+    // named after what it's called somewhere in the graph. It will help here to have both a mapping
+    // from module to name and a set of all the currently-used names.
+    var mod_names = std.AutoHashMap(*Module, []const u8).init(cs.builder.allocator);
+    var names = std.StringHashMap(void).init(cs.builder.allocator);
+
+    var to_name = std.ArrayList(struct {
+        name: []const u8,
+        mod: *Module,
+    }).init(cs.builder.allocator);
+    {
+        var it = cs.modules.iterator();
+        while (it.next()) |kv| {
+            // While we're traversing the root dependencies, let's make sure that no module names
+            // have colons in them, since the CLI forbids it. We handle this for transitive
+            // dependencies further down.
+            if (std.mem.indexOfScalar(u8, kv.key_ptr.*, ':') != null) {
+                @panic("Module names cannot contain colons");
+            }
```

Comment:
Could this check be done sooner so that it does not have to be done while traversing the graph? In other words, if every function that creates a module checks the name right away, then the error message will be more helpful, since the stack trace will lead directly to the problem, and then no need to look for colons while doing this unrelated thing.

Reply:
I did consider that. The problem is, currently to use any structure with dependency loops, you need to update a module's dependencies after creating it. We currently don't have any wrapper for that, so the API is basically just `mod.dependencies.put`, so we can't reliably add any validation at that level unless we (for instance) add a `Module.deoendOn` which does this check and discourage direct access to `dependencies`.

Reply:
Hmm, I see. I think this is fine then. Thanks for explaining.

Reply:
Cool - I'm planning another PR to solve the thing I mentioned on Discord where you can't currently test modules with dependency loops due to not being able to depend on the main module, I might revisit this then to at least try and improve the errors a bit.